### PR TITLE
Draft: Retry timed out requests from request

### PR DESCRIPTION
**Pull request by Gnuxie**
Hey, I'm not really sure what I'm doing. This is a first draft and I've gone with what I feel like is the least intrusive and simplest solution, I'll be really happy to change this if there is a better way to go about it.

My reason for submitting this PR is to make [Mjolnir](https://matrix.org/matrix-org/mjolnir) a little more resilient to longer requests in the bigger rooms. One thing that is important in this regard is being able to notify of a slow request when the request is the result of a command. We also think in Mjolnir's case some requests should fail sooner than later because of time sensitivity of some commands in their response to abuse. In this PR we emit an event on the MatrixClient's event emitter when a request is about to be retried so that we can handle both of these points.

I could have wrapped the request function and implemented it all from Mjolnir using `http.ts`'s `setRequestFn` but this would make handling the two constraints I mentioned a little bit less ergonomic and it would mean other user's of matrix-bot-sdk wouldn't be able to use the solution, if there is a demand for that? I will do this as a backup plan though.

Relevant but slightly out of date now: https://github.com/turt2live/matrix-bot-sdk/issues/18#issuecomment-990052384

### CODE REVIEW

File: src/MatrixClient.ts [hunk 1]
```diff
@@ -1880,10 +1885,62 @@ export class MatrixClient extends EventEmitter {
         if (this.accessToken) {
             headers["Authorization"] = `Bearer ${this.accessToken}`;
         }
-        return doHttpRequest(this.homeserverUrl, method, endpoint, qs, body, headers, timeout, raw, contentType, noEncoding);
+
+        const backoffTime = (nRetries: number) => {
+            return 1.6 ** nRetries * timeout;
```
Comment (Yoric) [inline src/MatrixClient.ts:1890]: Nit: Feels that this `1.6` should be a named constant (or a parameter).

File: src/MatrixClient.ts [hunk 2]
```diff
@@ -1880,10 +1885,62 @@ export class MatrixClient extends EventEmitter {
...
+                }
+            }
+        }
     }
 }
 
+/**
+ * This is used to identify requests from the client that are being retried.
+ */
+let clientRequestId = 0;
```
Comment (Yoric) [inline src/MatrixClient.ts:1942]: You're going to get race conditions reading this. You need to capture the value when you start `doRequest` if you wish to be consistent.

**Review by Yoric — CHANGES_REQUESTED** [review]
Note: I can't approve a PR on this project because I'm not a dev on it.

But I can make a first pass of review :)

**Comment by turt2live** [comment]
@Gnuxie is this still a draft, and are you able to take a look at the conflicts?

**Comment by turt2live** [comment]
Closing as the changes appear to be abandoned - please open a new PR if needed.

**Pull request closed without merging**